### ISSUE TITLE
Need parameter for generators other than pythia

### ISSUE THREAD

**Issue by ericvaandering** [dmwm/CRABClient#4073]
See  https://hypernews.cern.ch/HyperNews/CMS/get/crabDevelopment/1975/1/1/1.html for context. Will also require a ticket against CRABServer to reconfigure CMSSW.

**Comment by AndresTanasijczuk** [comment]
Can we consider this ticket as resolved? 
We have now the "generator" parameter in CRAB3, but I think it only works for lhe and not for other generators... @matz-e ? 
If it only works for lhe, what is the priority level for extending to other generators (which ones? comphep? madgraph?)? @ericvaandering @belforte

**Comment by belforte** [comment]
My memory of the conversation with Josh is that if things work for Pythia and LHE, then work for everything. There are only two classes of generators now, IIUC, LHE and everything else. But am not familiar with generator parameter in Crab3. It may be that you only need to review/update he documentation. What does it exactly do ?

**Comment by matz-e** [comment]
The `generator` parameter currently allows for only one value, `lhe`, and is optional.  I think that for clarity, we could opt to make it mandatory and permit more different values.  I would guess that other generation needs are met by the Pythia-equivalent mode of running?

**Comment by belforte** [comment]
I see potential for confusion if we encourage users to write a generator name string which we do not 
use other then "if == 'lhe'". Let's start by listing the different task types in terms of
what happens, and in parallel, how they are configured. That small table may shed light.

**Comment by ericvaandering** [comment]
I don’t understand this comment of making it mandatory. Analysis of data will not have a generator. On the other hand, we already support two generators, AFAIK: LHE files and Pythia.

Eric

> On Jan 8, 2015, at 09:05, Matthias Wolf <notifications@github.com> wrote:
> 
> The generator parameter currently allows for only one value, lhe, and is optional. I think that for clarity, we could opt to make it mandatory and permit more different values. I would guess that other generation needs are met by the Pythia-equivalent mode of running?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/dmwm/CRABClient/issues/4073#issuecomment-69191114>.
> 




Eric W. Vaandering 
Computing Physics Developer
CMS Computing Development Lead

Scientific Computing Division
Fermi National Accelerator Laboratory
630 962 7260 office
www.fnal.gov <http://www.fnal.gov/>

**Comment by belforte** [comment]
LHE is a file format which can be produced by many generators.
I fear we are diverging into "find some line to fix and close" rather
then "make sure we understand what we want to do and why".
Let's have plans before issues :-)

**Comment by matz-e** [comment]
@ericvaandering I was talking about making it mandatory for MC generation only.  It does not seem to be that way in CRAB2. To quote the man page of CRAB2:

```
generator

Name of the generator your MC job is using. Some generators require CRAB to skip events, others do not. Possible values are pythia (default), comphep, lhe, and madgraph. This will skip events in your generator input file.
```

**Comment by ericvaandering** [comment]
Right. In CRAB2 it defaults to pythia. But it can be made mandatory here if you think having a default is not sensible. For CRAB2 for a long, long time it only did pythia, so that’s the reasoning for the default.

Eric

> On Jan 8, 2015, at 10:19, Matthias Wolf <notifications@github.com> wrote:
> 
> @ericvaandering <https://github.com/ericvaandering> I was talking about making it mandatory for MC generation only. It does not seem to be that way in CRAB2. To quote the man page of CRAB2:
> 
> generator
> 
> Name of the generator your MC job is using. Some generators require CRAB to skip events, others do not. Possible values are pythia (default), comphep, lhe, and madgraph. This will skip events in your generator input file.
> —
> Reply to this email directly or view it on GitHub <https://github.com/dmwm/CRABClient/issues/4073#issuecomment-69203457>.
> 




Eric W. Vaandering 
Computing Physics Developer
CMS Computing Development Lead

Scientific Computing Division
Fermi National Accelerator Laboratory
630 962 7260 office
www.fnal.gov <http://www.fnal.gov/>

**Comment by belforte** [comment]
let's not take CRAB2 as example on this. Doing MC generation there has always been a dark art with 
no good documentation. Besides, Josh claimed that things are much improved now from CMSSW side, so 
let's try to catch up !

**Comment by ericvaandering** [comment]
Right. I think what that means is that mad graph and comp hep etc are moved into LHE. 



> On Jan 8, 2015, at 10:30, Stefano Belforte <notifications@github.com> wrote:
> 
> let's not take CRAB2 as example on this. Doing MC generation there has always been a dark art with 
> no good documentation. Besides, Josh claimed that things are much improved now from CMSSW side, so 
> let's try to catch up !
> —
> Reply to this email directly or view it on GitHub <https://github.com/dmwm/CRABClient/issues/4073#issuecomment-69205263>.
> 




Eric W. Vaandering 
Computing Physics Developer
CMS Computing Development Lead

Scientific Computing Division
Fermi National Accelerator Laboratory
630 962 7260 office
www.fnal.gov <http://www.fnal.gov/>

**Comment by AndresTanasijczuk** [comment]
Can we close this issue now? @matz-e has introduced a default value ('pythia') for the generator parameter.

**Comment by ericvaandering** [comment]
Yes, I think so. New generators (if any) will be a new ticket.
